File: python/knight/stream.py
import re
from typing import Union 

class Stream():
	""" The class used when parsing data. """
	WHITESPACE: re.Pattern = re.compile(r'([\s(){}\[\]:]+|\#[^\n]*)+')

	def __init__(self, source: str):
		""" Creates a new `Stream` with the given source. """
		self.source = source

	def __str__(self) -> str:
		""" Returns the remainder of the stream. """
		return self.source

	def __bool__(self) -> bool:
		""" Returns whether the stream is empty. """
		return bool(self.source)

	def strip(self):
		""" Removes all leading whitespace and quotes """
		self.matches(Stream.WHITESPACE)

	def peek(self) -> Union[None, str]:
		""" Returns the first character of the stream """
		return self.source[0] if self.source else None

	def matches(self, rxp: re.Pattern, index: int = 0) -> Union[None, str]:
		"""
		Checks to see if the start of the stream matches `rxp`.

		If the stream doesn't match, `None` is returned. Otherwise, the
		stream is updated, and the `index`th group is returned. (The
		default value of `0` means the entire matched regex is returned.)
		"""
		match = rxp.match(self.source)

		if match is None:
			return None

		string = self.source[:match.end()]
		self.source = self.source[match.end():]
		return match[index]

File: php/src/Stream.php
<?php
namespace Knight;

/**
 * A Stream that's used when parsing Values.
 **/
class Stream
{
	/**
	 * The stream of text that's to be parsed.
	 *
	 * @var string
	 **/
	private $stream;

	/**
	 * Creates a new Stream for the given $source
	 *
	 * @var string $source The string to use as the source for this Stream.
	 **/
	public function __construct(string $source)
	{
		$this->source = $source;
	}

	/**
	 * Removes all leading whitespace and comments.
	 *
	 * Note that, for Knight, all forms of parens (ie `[`, `]`, `(`, `)`, `{`, and `}`), as well as the colon (`:`) are
	 * considered whitespace.
	 *
	 * @return void
	 **/
	public function strip(): void
	{
		$this->source = preg_replace('/\A(?:[\]\[\s(){}:]+|\#[^\n]*(\n|$))*/m', '', $this->source);
	}

	/**
	 * Attempts to match the $regex at the start of the source.
	 *
	 * If the regex matches, the entire matching string will be returned by default. The `$idx` parameter can be used to
	 * change this behaviour around.
	 *
	 * @param string $regex The regex to match at the beginning of the string; It will have the `/m` flag applied.
	 * @param int $idx The index of the group to return; defaults to `0`, ie the entire match.
	 * @return string|null Returns the matching string (or the `$idx` capture group) if `$regex` matched. Else, null.
	 **/
	public function match(string $regex, int $idx=0): ?string
	{
		if (!preg_match("/\A(?:$regex)/m", $this->source, $match)) {
			return null;
		}

		$this->source = substr($this->source, strlen($match[0]));

		return $match[$idx];
	}
}
